File: src/tools/registry.ts
import type { AwsProvider } from "../integrations/aws/provider.js";
import type { ToolResult } from "../schemas/agent.js";
import {
  queryAiopsSummary,
  queryCostAnomalies,
  queryIncidentDiagnosis,
  queryRunbookRecommendations,
} from "./aiops.js";
import { assertReadOnlyTool, ReadOnlyViolation } from "./guard.js";
import {
  queryAlerts,
  queryClusterStatus,
  queryLogs,
  queryResourceInventory,
  queryServiceCost,
} from "./platform.js";

export type ToolFunction = (provider: AwsProvider) => Promise<ToolResult>;

export class UnknownToolError extends Error {}

export class ToolRegistry {
  private readonly tools: Record<string, ToolFunction>;

  constructor(private readonly provider: AwsProvider) {
    this.tools = {
      query_alerts: queryAlerts,
      query_service_cost: queryServiceCost,
      query_cluster_status: queryClusterStatus,
      query_resource_inventory: queryResourceInventory,
      query_logs: queryLogs,
      query_incident_diagnosis: queryIncidentDiagnosis,
      query_cost_anomalies: queryCostAnomalies,
      query_runbook_recommendations: queryRunbookRecommendations,
      query_aiops_summary: queryAiopsSummary,
    };

    for (const name of this.names()) {
      assertReadOnlyTool(name);
    }
  }

  names(): string[] {
    return Object.keys(this.tools);
  }

  selectForMessage(message: string): string[] {
    const text = message.toLowerCase();
    const selectionRules: Record<string, string[]> = {
      query_alerts: ["alert", "alarm", "incident", "告警", "报警"],
      query_service_cost: ["cost", "finops", "spend", "bill", "费用", "成本", "账单"],
      query_cluster_status: ["eks", "cluster", "pod", "node", "集群", "节点"],
      query_resource_inventory: ["resource", "inventory", "ec2", "rds", "s3", "资源"],
      query_logs: ["log", "error", "trace", "日志", "错误"],
      query_incident_diagnosis: ["diagnose", "incident", "root cause", "rca", "故障", "诊断", "根因"],
      query_cost_anomalies: ["anomaly", "cost spike", "overspend", "异常", "费用异常"],
      query_runbook_recommendations: ["runbook", "triage", "next step", "预案", "下一步"],
      query_aiops_summary: ["aiops", "overview", "summary", "all signals", "全套", "总结", "总览"],
    };

    return Object.entries(selectionRules)
      .filter(([, keywords]) => keywords.some((keyword) => text.includes(keyword)))
      .map(([name]) => name);
  }

  async run(name: string): Promise<ToolResult> {
    assertReadOnlyTool(name);
    const tool = this.tools[name];
    if (!tool) {
      throw new UnknownToolError(`Unknown tool '${name}'.`);
    }
    return tool(this.provider);
  }
}

export function isReadOnlyViolation(error: unknown): error is ReadOnlyViolation {
  return error instanceof ReadOnlyViolation;
}

File: tools/registry.py
from collections.abc import Awaitable, Callable
from typing import Any

from fastapi import HTTPException

from integrations.aws.provider import AwsProvider, get_aws_provider
from tools.alerts import query_alerts
from tools.aiops import (
    query_aiops_summary,
    query_cost_anomalies,
    query_incident_diagnosis,
    query_runbook_recommendations,
)
from tools.eks import query_cluster_status
from tools.finops import query_service_cost
from tools.guard import ReadOnlyViolation, assert_read_only_tool
from tools.logs import query_logs
from tools.resources import query_resource_inventory

ToolFunc = Callable[[AwsProvider], Awaitable[dict[str, Any]]]


class ToolRegistry:
    def __init__(self, provider: AwsProvider) -> None:
        self.provider = provider
        self._tools: dict[str, ToolFunc] = {
            "query_alerts": query_alerts,
            "query_service_cost": query_service_cost,
            "query_cluster_status": query_cluster_status,
            "query_resource_inventory": query_resource_inventory,
            "query_logs": query_logs,
            "query_incident_diagnosis": query_incident_diagnosis,
            "query_cost_anomalies": query_cost_anomalies,
            "query_runbook_recommendations": query_runbook_recommendations,
            "query_aiops_summary": query_aiops_summary,
        }
        for name in self._tools:
            assert_read_only_tool(name)

    def names(self) -> list[str]:
        return list(self._tools.keys())

    def select_for_message(self, message: str) -> list[str]:
        text = message.lower()
        selection_rules = {
            "query_alerts": ["alert", "alarm", "incident"],
            "query_service_cost": ["cost", "finops", "spend", "bill"],
            "query_cluster_status": ["eks", "cluster", "pod", "node"],
            "query_resource_inventory": ["resource", "inventory", "ec2", "rds", "s3"],
            "query_logs": ["log", "error", "trace"],
            "query_incident_diagnosis": ["diagnose", "incident", "root cause", "rca", "degradation"],
            "query_cost_anomalies": ["anomaly", "cost spike", "overspend", "finops"],
            "query_runbook_recommendations": ["runbook", "triage", "next step"],
            "query_aiops_summary": ["aiops", "overview", "summary", "all signals", "full system"],
        }
        return [
            name
            for name, keywords in selection_rules.items()
            if any(keyword in text for keyword in keywords)
        ]

    async def run(self, name: str) -> dict[str, Any]:
        try:
            assert_read_only_tool(name)
        except ReadOnlyViolation as exc:
            raise HTTPException(status_code=403, detail=str(exc)) from exc

        tool = self._tools.get(name)
        if tool is None:
            raise HTTPException(status_code=404, detail=f"Unknown tool '{name}'.")
        return await tool(self.provider)


def get_tool_registry() -> ToolRegistry:
    return ToolRegistry(provider=get_aws_provider())
